### PR TITLE
o/ifacestate{,apparmorprompting}: do not hold state lock during prompting startup

### PR DESCRIPTION
**Pull request by olivercalder**
Recording notices requires state lock, but recording notices asynchronously can cause unexpected behavior for the client. Thus, explicitly unlock state if the state lock is held when we know notices may be recorded; namely, during ifacemgr `StartUp`.

This reverts #14405. This is tracked internally by https://warthogs.atlassian.net/browse/SNAPDENG-31228

### CODE REVIEW

File: overlord/ifacestate/ifacemgr.go [hunk 1]
```diff
@@ -273,6 +279,9 @@ func (m *InterfaceManager) Ensure() error {
 // if running.
 func (m *InterfaceManager) Stop() {
 	m.stopUDevMon()
+	// XXX: The state lock is *not* held during Stop(). If it were, we'd need
+	// to unlock state and re-lock it after the call to
+	// interfacesRequestsManagerStop, so that it can record notices if needed.
```
Comment (olivercalder) [inline overlord/ifacestate/ifacemgr.go:284]: @pedronis please confirm that this is the case, from what I could tell this is true but not certain

Comment (pedronis) [reply overlord/ifacestate/ifacemgr.go:284]: yes, afaik we don't hold the state lock when entering any of the manager methods driven by StateEngine, you could probably replace the XXX with a comment about that

**Review by pedronis — COMMENTED** [review]
I would suggest to also do this in the tests:

```
diff --git a/overlord/ifacestate/ifacestate_test.go b/overlord/ifacestate/ifacestate_test.go
index 6836cec88e..aa808bc362 100644
--- a/overlord/ifacestate/ifacestate_test.go
+++ b/overlord/ifacestate/ifacestate_test.go
@@ -378,12 +378,16 @@ func (s *interfaceManagerSuite) TestSmokeAppArmorPromptingEnabled(c *C) {
        fakeManager := &apparmorprompting.InterfacesRequestsManager{}
        restore = ifacestate.MockCreateInterfacesRequestsManager(func(s *state.State) (*apparmorprompting.InterfacesRequestsManager, error) {
                createCount++
+               s.Lock()
+               s.Unlock()
                return fakeManager, nil
        })
        defer restore()
        stopCount := 0
        restore = ifacestate.MockInterfacesRequestsManagerStop(func(m *apparmorprompting.InterfacesRequestsManager) error {
                stopCount++
+               s.st.Lock()
+               s.st.Unlock()
                return nil
        })
        defer restore()
```

Maybe there are other or better tests to do that in?

Also as suggested elsewhere we should have stop - start test for snapd with the apparmor  prompting feature kept enabled

**Review by pedronis — APPROVED** [review]
thank you

**Comment by olivercalder** [comment]
Had a typo :(

**Comment by olivercalder** [comment]
Client integration tests are passing again with this PR.

**Review by ernestl — APPROVED** [review]
Thanks

**Comment by olivercalder** [comment]
This PR is ready to be merged when ready @ernestl . The core24 failure is main degraded, just flaky I believe, and the noble failure is `upgrade-from-release`, which is unrelated and fixed by https://github.com/canonical/snapd/pull/14416

**Comment by olivercalder** [comment]
I don't believe recording notices in goroutines is actually incorrect, it I think it just may be confusing to the test runtime. Some investigation here: https://github.com/canonical/prompting-client/pull/81

Regardless, I think it's clearer for the client if all notices occur in order and before API calls return, so this PR is still worth merging. In the future, we can explore ways to record notices synchronously without requiring the state lock.

**Comment by olivercalder** [comment]
I rebased to clean up the history and preserve the `revert` commit.

**Pull request merged**